### PR TITLE
[WIP] Equinix Metal Support

### PR DESCRIPTION
**Pull request by detiber**
Depends on:
https://github.com/openshift/cloud-credential-operator/pull/283
https://github.com/openshift/api/pull/784

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -39,3 +39,5 @@ require (
 replace sigs.k8s.io/cluster-api-provider-aws => github.com/openshift/cluster-api-provider-aws v0.2.1-0.20201022175424-d30c7a274820
 
 replace sigs.k8s.io/cluster-api-provider-azure => github.com/openshift/cluster-api-provider-azure v0.1.0-alpha.3.0.20201016155852-4090a6970205
+
+replace github.com/openshift/api => github.com/detiber/api v0.0.0-20201113192824-c19de3594400
```
Comment (detiber) [inline go.mod:43]: Relies on changes in https://github.com/openshift/api/pull/784

**Review by elmiko — COMMENTED** [review]
this looks reasonable to me. i have not had a chance to run it yet, although the binary build and unit tests worked. i had a minor suggestion about the CCO repo, but it's not a blocker.

i think we should put a hold here just to make sure we don't do anything before the api changes merge.
/lgtm
/hold until openshift/api#784 has merged

**Comment by elmiko** [comment]
one more comment, this should be rebased as some of the baremetal stuff has been removed since it was first posted.

File: install/0000_30_machine-api-operator_00_credentials-request.yaml [hunk 2]
```diff
@@ -166,3 +166,18 @@ spec:
...
+  name: openshift-machine-api-equinix-metal
+  namespace: openshift-cloud-credential-operator
+spec:
+  secretRef:
+    name: equinix-metal-credentials
+    namespace: openshift-machine-api
+  providerSpec:
+    apiVersion: cloudcredential.openshift.io/v1
+    kind: EquinixMetalProviderSpec
+---
```
[the remark this answers is not in the document]
Comment (detiber) [reply install/0000_30_machine-api-operator_00_credentials-request.yaml:183]: Done, created: https://github.com/openshift/cloud-credential-operator/pull/283

File: go.mod [hunk 3]
```diff
@@ -37,3 +37,5 @@ require (
 replace sigs.k8s.io/cluster-api-provider-aws => github.com/openshift/cluster-api-provider-aws v0.2.1-0.20201125052318-b85a18cbf338
 
 replace sigs.k8s.io/cluster-api-provider-azure => github.com/openshift/cluster-api-provider-azure v0.1.0-alpha.3.0.20201130182513-88b90230f2a4
+
+replace github.com/openshift/api => github.com/detiber/api v0.0.0-20210113181726-19e9a251beff
```
Comment (detiber) [inline go.mod:41]: Temporary workaround until https://github.com/openshift/api/pull/784 is merged

**Comment by detiber** [comment]
Rebased, I'm having difficulty running the tests locally, they don't like running outside of docker (running hack/ci-test.sh directly due to not having the kubebuilder binaries installed where it wants them) and running in docker doesn't seem happy. 

Not sure how to reconcile at the moment, but may try digging in further a little be latter on down the line.

File: go.mod (hunk 3 above)
Comment (elmiko) [reply go.mod:41]: +1, that makes sense for now

**Comment by elmiko** [comment]
> Rebased, I'm having difficulty running the tests locally, they don't like running outside of docker (running hack/ci-test.sh directly due to not having the kubebuilder binaries installed where it wants them) and running in docker doesn't seem happy.

yeah, i apologize for this but it's a known deficiency currently. i will spend some time to see if i can provide a good recipe for running these.

> Not sure how to reconcile at the moment, but may try digging in further a little be latter on down the line.

i am going to take another look as well to see if we can't make this easier for you. this is a known issue

**Comment by elmiko** [comment]
@detiber i just saw that this pr got posted https://github.com/openshift/machine-api-operator/pull/789 , i haven't tested it too much yet but just wanted to share.

**Comment by detiber** [comment]
Updated, did some squashing, added the commit from #789 to enable local testing

**Comment by elmiko** [comment]
those changes worked for me, i was able to build the binary with `make build` and run the unit tests with `make test`. bonus, all the tests passed.

**Comment by elmiko** [comment]
just wanted to leave an update here, i am still working towards live testing of the equinix metal provider. i have access to the equinix platform and have been working through the upi installation process for openshift. my next step will be to add in the equinix specific provider bits and run some manual testing followed by attempting to run our e2e tests.

**Pull request closed without merging**